### PR TITLE
Only broadcast `HostRemovedFromCluster` when a host is part of a cluster

### PR DESCRIPTION
**Pull request by jamie-suse**
Description
===========

Relates to https://github.com/trento-project/web/pull/1474

> This PR fixes possible concurrency issues when a host is removed from a cluster (delta-deregistration flow).
> The issue is caused by two aggregates acting concurrently, so the HostAddedToCluster event and HostRemovedFromCluster event ordering is not guaranteed.
>
> The solution is to change the projector logic to skip the event if the host is not part of the cluster anymore.
>
> **We still need to conditionally broadcast to the FE the event, but this will come once the FE PR is merged.**

This change will broadcast `HostRemovedFromCluster` _only_ when a given host is part of a cluster.

How was this tested?
--------------------

Added conditions in unit tests to assert for the existence/nonexistence of the broadcast messages when `HostRemovedFromCluster` is called

### CODE REVIEW

**Review by arbulu89 — CHANGES_REQUESTED** [review]

File: lib/trento/application/projectors/host_projector.ex [hunk 1]
```diff
@@ -275,6 +275,16 @@ defmodule Trento.HostProjector do
     end
   end
 
+  def after_update(
+        %HostRemovedFromCluster{host_id: host_id},
+        _,
+        %{host: %Trento.HostReadModel{cluster_id: nil}}
+      ) do
+    TrentoWeb.Endpoint.broadcast("monitoring:hosts", "host_details_updated", %{
+      id: host_id
```
Comment (jamie-suse) [inline lib/trento/application/projectors/host_projector.ex:284]: Is this the correct payload to broadcast? `cluster_id` will always be `nil` because of the pattern-matching, is it useful to transmit this?

File: test/trento/application/projectors/host_projector_test.exs [hunk 2]
```diff
@@ -148,8 +148,12 @@ defmodule Trento.HostProjectorTest do
     }
 
     ProjectorTestHelper.project(HostProjector, event, "host_projector")
-    projection = Repo.get!(HostReadModel, host_id)
 
+    assert_broadcast "host_details_updated",
```
Comment (arbulu89) [inline test/trento/application/projectors/host_projector_test.exs:152]: Why did you change the order between the `assert_broadcast` and `Repo.get`?
Every other test has the `assert_broadcast` as last thing in the test

File: test/trento/application/projectors/host_projector_test.exs [hunk 3]
```diff
@@ -172,6 +176,10 @@ defmodule Trento.HostProjectorTest do
     projection = Repo.get!(HostReadModel, host_id)
 
     assert cluster_id == projection.cluster_id
+
+    refute_broadcast "host_details_updated",
+                     %{id: ^host_id},
+                     1000
```
Comment (arbulu89) [inline test/trento/application/projectors/host_projector_test.exs:182]: I'm just a bit worried about this `1000` value.
In the `refute` this blocks the test until the timeout is expired, so it will be waiting for 1 second to have a green test.
Maybe we should use the default `100` value?
@CDimonaco @fabriziosestito Any suggestion on this?

**Review by arbulu89 — APPROVED** [review]
All good from my side!
Only commented the `refute_broadcast` timeout value, which slows down or tests

File: test/trento/application/projectors/host_projector_test.exs (hunk 3 above)
Comment (CDimonaco) [reply test/trento/application/projectors/host_projector_test.exs:182]: I think the 1000 value was set by me when I started doing broadcast test, this was set by testing manually different values, if the default timeout is sufficient we can switch to that

**Review by CDimonaco — APPROVED** [review]

File: test/trento/application/projectors/host_projector_test.exs (hunk 3 above)
Comment (arbulu89) [reply test/trento/application/projectors/host_projector_test.exs:182]: I mean, in `assert_broadcast` is fine, because it will just wait until the message is sent. And it will be a long test only on failure.
The `refute` will always wait for the time, even in passing tests.
That's the difference

Comment (jamie-suse) [reply test/trento/application/projectors/host_projector_test.exs]: I reduced this timeout to 100 mS 👍

Comment (arbulu89) [reply test/trento/application/projectors/host_projector_test.exs]: Just a least request XD
Could you test if putting this same `refute_broadcast` with 100ms in the test above fails? Of course, just temporarily, in your local env.
It should fail. I want to be sure that the 100ms makes sense

Comment (jamie-suse) [reply test/trento/application/projectors/host_projector_test.exs]: Just did on my local machine and `refute_broadcast` can catch messages using a `timeout` as low as 1 mS 😵 I guess we may see different performance in the CI environment though.

**Comment by arbulu89** [comment]
@jamie-suse Green light from my side!
